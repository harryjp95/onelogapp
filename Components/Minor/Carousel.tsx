import fetch from 'node-fetch'
import React, { ReactElement, useEffect, useState } from 'react'
import { ActivityIndicator, ScrollView, StyleSheet, View } from 'react-native'

import { GradientBorder } from './GradientBorder'

interface CarouselProps {
  table: string,
  Template: ReactElement
}

const getReadings = (table: string) => {
  const url = `http://localhost:8088/readings/${table}`

  return fetch(url, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    }
  })
    .then(res => res.json())
    .catch(err => err)
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { table, Template } = props
  const [readings, setReadings] = useState()

  useEffect(() => {
    getReadings(table).then(res => setReadings(res.reverse()))
  }, [])

  let templates
  if (readings) {
    templates = readings.map((r, i) => {
      return <Template created={r.created} reading={r.reading} index={i} />
    })
  }

  return(
    readings
      ? <>
        <ScrollView horizontal={true} pagingEnabled={true}>
          { templates }
        </ScrollView>
        <GradientBorder x={1.0} y={1.0} />
        </>
      : <View style={{ flex: 1 }}>
          <View style={{ flex: 1 }} />
          <View style={{ ...Styles.lastReading }}>
            <ActivityIndicator color={'black'} />
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
  )
}

const Styles = StyleSheet.create({
  lastReading: {
    flex: 5,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  }
})

export default Carousel
