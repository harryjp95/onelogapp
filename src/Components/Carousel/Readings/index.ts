export * as BgCarousel from './Bg'
export * as StatsCarousel from './Stats'
export * as DoseCarousel from './Dose'
export * as MacroCarousel from './Macro'
