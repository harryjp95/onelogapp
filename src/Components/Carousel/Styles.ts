import { StyleSheet } from 'react-native'

import Colors from '../../Assets/Styles/Colors'

export const CarouselStyles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.lightGrey1
  },
  header: {
    flex: 1.1,
    flexDirection: 'row',
    alignItems: 'center',
    width: '100%'
  },
  contentContainer: {
    flexDirection: 'row',
    flex: 5,
    alignItems: 'center',
    justifyContent: 'center'
  },
  template: {
    flex: 4,
    alignContent: 'center'
  },
  chevron: {
    flex: 1
  },
  tag: {
    fontSize: 14,
    fontWeight: 'bold',
    textAlign: 'center',
    borderRightWidth: 1,
    padding: 4,
    width: 60
  },
  time: {
    width: '100%',
    fontSize: 16,
    fontWeight: 'bold',
    padding: 4
  },
  lastReading: {
    flex: 5,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  },
  placeholder: {
    flex: 1
  }
})

export const BgStyles = StyleSheet.create({
  container: {
    flex: 5,
    paddingBottom: 10,
    backgroundColor: Colors.lightGrey1,
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center'
  },
  reading: {
    justifyContent: 'center',
    color: Colors.black,
    fontSize: 54,
    paddingTop: 8
  },
  unit: {
    fontSize: 12,
    paddingBottom: 10
  },
  image: {
    height: 30,
    width: 30
  }
})

export const DoseStyles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    width: '100%'
  },
  readingContainer: {
    alignItems: 'center',
    width: '50%'
  },
  reading: {
    fontSize: 54,
    textAlign: 'center',
    paddingTop: 8,
    color: Colors.black
  },
  unit: {
    fontSize: 12
  },
  colorHighlight: {
    width: '100%'
  }
})

export const MacroStyles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    backgroundColor: Colors.lightGrey1,
    width: '100%'
  },
  labels: {
    flexDirection: 'column',
    padding: 20
  },
  label: {
    fontSize: 16,
    color: Colors.darkGrey2
  },
  values: {
    flexDirection: 'column',
    padding: 20
  },
  value: {
    fontSize: 16,
    fontWeight: 'bold',
    alignSelf: 'flex-end',
    color: Colors.black
  }
})

export const StatsStyles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    width: '100%'
  },
  readingContainer: {
    alignItems: 'center',
    width: '50%'
  },
  reading: {
    fontSize: 54,
    color: Colors.black
  },
  unit: {
    fontSize: 12
  },
  stddev: {
    textAlign: 'center',
    fontSize: 20,
    fontWeight: 'bold',
    padding: 12,
    borderWidth: 1,
    borderRadius: 4,
    borderColor: Colors.lightGrey4,
    width: '30%'
  }
})
