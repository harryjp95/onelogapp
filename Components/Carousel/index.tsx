import React, { useState, ReactElement } from 'react'
import { ActivityIndicator, StyleSheet, Text, View } from 'react-native'
import GestureRecognizer from 'react-native-swipe-gestures';

import { GradientBorder } from '../Minor/GradientBorder'
import { Chevron } from '../Minor/Chevron'
import { generateCreatedDate } from '../Helpers/DateHelpers'
import { capitalise } from '../Helpers/GeneralHelpers'

interface CarouselProps {
  table: string,
  Template: React.FC,
  readings: ReactElement[]
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { table, Template, readings } = props
  const [index, setIndex] = useState(0)

  const handleSwipeLeft = () => {
    if (index < readings.length - 1) setIndex(index + 1)
  }

  const handleSwipeRight = () => {
    if (index > 0) setIndex(index - 1)
  }

  const reading = readings && readings[index]

  return(
    readings.length > 0
      ? <View style={Styles.container}>
          <View style={Styles.header}>
            <Text style={Styles.tag}>
              { capitalise(table) }
            </Text>
            <Text style={Styles.time}>
              { generateCreatedDate(reading['created']) }
            </Text>
          </View>
          <GradientBorder x={0.4} y={1.0} colors={['grey', '#ebebeb']} />

          <View style={Styles.contentContainer}>
            <View style={Styles.chevron}>
            {index === 0
              ? <Chevron symbol={''} handlePress={() => null}/>
              : <Chevron symbol={'<'} handlePress={handleSwipeRight} />}
            </View>
            <GestureRecognizer onSwipeLeft={handleSwipeLeft} onSwipeRight={handleSwipeRight} style={Styles.template}>
              <Template data={reading} />
            </GestureRecognizer>
            <View style={Styles.chevron}>
            {index < readings.length - 1
              ? <Chevron symbol={'>'} handlePress={handleSwipeLeft} />
              : <Chevron symbol={''} handlePress={() => null} />}
            </View>
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
      : <View style={{ flex: 1 }}>
          <View style={{ flex: 1 }} />
          <View style={{ ...Styles.lastReading }}>
            <ActivityIndicator color={'black'} />
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
  )
}

const Styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#ebebeb'
  },
  header: {
    flex: 1,
    flexDirection: 'row',
    width: '100%',
  },
  contentContainer: {
    flexDirection: 'row',
    flex: 5,
    alignItems: 'center',
    justifyContent: 'center'
},
  template: {
    flex: 4,
    alignContent: 'center'
  },
  chevron: {
    flex: 1
  },
  tag: {
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
    padding: 4,
    color: 'white',
    backgroundColor: 'black',
    width: 60
  },
  time: {
    fontSize: 16,
    fontWeight: 'bold',
    padding: 4
  },
  lastReading: {
    flex: 5,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  }
})

export default Carousel
