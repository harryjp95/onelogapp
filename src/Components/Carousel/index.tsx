import React, { useEffect, useState } from 'react'
import { ActivityIndicator, StyleSheet, Text, View } from 'react-native'
import GestureRecognizer from 'react-native-swipe-gestures';

import { IBgReading } from './Readings/Bg'
import { IStatsReading } from './Readings/Stats'
import { IDoseReading } from './Readings/Dose'
import { IMacroReading } from './Readings/Macro';

import { getData } from '../Store/index'
import { GradientBorder } from '../Minor/GradientBorder'
import { Chevron } from '../Minor/Chevron'
import { generateCreatedDate } from '../Helpers/Date'
import { capitalise } from '../Helpers/General'

export interface BgTemplateProps {
  data: IBgReading
}
export interface StatsTemplateProps {
  data: IStatsReading
}
export interface DoseTemplateProps {
  data: IDoseReading
}
export interface MacroTemplateProps {
  data: IMacroReading
}

interface CarouselProps {
  table: string,
  Template: React.FC<BgTemplateProps> | React.FC<StatsTemplateProps> | React.FC<DoseTemplateProps> | React.FC<MacroTemplateProps>,
  dataKey: string
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { table, Template, dataKey } = props
  const [readings, setReadings] = useState([] as any)
  const [index, setIndex] = useState(0)

  useEffect(() => {
    const fetchReadings = async (key: string) => {
      try {
        const data = await getData(key)
        if (data && data.readings) {
          setReadings(data.readings)
        }
      } catch(err) {
        console.log('Error fetchReadings: ', err)
      }
    }
    fetchReadings(dataKey)
  }, [])

  const handleSwipeLeft = () => {
    if (index < readings.length - 1) setIndex(index + 1)
  }

  const handleSwipeRight = () => {
    if (index > 0) setIndex(index - 1)
  }

  const reading = readings && readings[index]

  return(
    readings.length > 0
      ? <View style={Styles.container}>
          <View style={Styles.header}>
            <Text style={Styles.tag}>
              { capitalise(table) }
            </Text>
            <Text style={Styles.time}>
              { generateCreatedDate(reading['created']) }
            </Text>
          </View>
          <GradientBorder x={0.4} y={1.0} colors={['grey', '#ebebeb']} />

          <View style={Styles.contentContainer}>
            <View style={Styles.chevron}>
            {index === 0
              ? <Chevron symbol={''} handlePress={() => null}/>
              : <Chevron symbol={'<'} handlePress={handleSwipeRight} />}
            </View>
            <GestureRecognizer onSwipeLeft={handleSwipeLeft} onSwipeRight={handleSwipeRight} style={Styles.template}>
              <Template data={reading} />
            </GestureRecognizer>
            <View style={Styles.chevron}>
            {index < readings.length - 1
              ? <Chevron symbol={'>'} handlePress={handleSwipeLeft} />
              : <Chevron symbol={''} handlePress={() => null} />}
            </View>
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
      : <View style={{ flex: 1 }}>
          <View style={{ flex: 1 }} />
          <View style={{ ...Styles.lastReading }}>
            <ActivityIndicator color={'black'} />
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
  )
}

const Styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#ebebeb'
  },
  header: {
    flex: 1,
    flexDirection: 'row',
    width: '100%',
  },
  contentContainer: {
    flexDirection: 'row',
    flex: 5,
    alignItems: 'center',
    justifyContent: 'center'
},
  template: {
    flex: 4,
    alignContent: 'center'
  },
  chevron: {
    flex: 1
  },
  tag: {
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
    padding: 4,
    color: 'white',
    backgroundColor: 'black',
    width: 60
  },
  time: {
    fontSize: 16,
    fontWeight: 'bold',
    padding: 4
  },
  lastReading: {
    flex: 5,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  }
})

export default Carousel
