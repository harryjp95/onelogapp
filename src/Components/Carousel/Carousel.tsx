import React, { useState } from 'react'
import { ActivityIndicator, Text, View } from 'react-native'
import GestureRecognizer from 'react-native-swipe-gestures'

import { Chevron, GradientBorder } from '../Minor'
import { capitalise, generateCreatedDate } from '../../Helpers'
import { Reading } from '../../types'
import { CarouselStyles } from './Styles'

interface ReadingCarouselProps {
  reading: Reading
}

interface CarouselProps {
  name: string
  Template: React.FC<ReadingCarouselProps>
  readings: any[]
  startingIndex?: number
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { name, Template, readings, startingIndex } = props
  const [index, setIndex] = useState(startingIndex || 0)

  const handleSwipeLeft = () => {
    if (index < readings.length - 1) setIndex(index + 1)
  }

  const handleSwipeRight = () => {
    if (index > 0) setIndex(index - 1)
  }

  const reading = readings && readings[index]

  return readings && readings.length > 0 ? (
    <View style={CarouselStyles.container} testID="carousel">
      <View style={CarouselStyles.header}>
        <Text style={CarouselStyles.tag}>{capitalise(name)}</Text>
        <Text style={CarouselStyles.time}>{generateCreatedDate(reading.created) || ''}</Text>
      </View>
      <GradientBorder x={0.4} y={1.0} colors={['grey', '#ebebeb']} />

      <View style={CarouselStyles.contentContainer}>
        <View style={CarouselStyles.chevron}>
          {index === 0 ? <Chevron handlePress={() => null} /> : <Chevron left handlePress={handleSwipeRight} />}
        </View>
        <GestureRecognizer onSwipeLeft={handleSwipeLeft} onSwipeRight={handleSwipeRight} style={CarouselStyles.template}>
          <Template reading={reading} />
        </GestureRecognizer>
        <View style={CarouselStyles.chevron}>
          {index < readings.length - 1 ? (
            <Chevron right handlePress={handleSwipeLeft} />
          ) : (
            <Chevron handlePress={() => null} />
          )}
        </View>
      </View>
      <GradientBorder x={1.0} y={1.0} />
    </View>
  ) : (
    <View style={CarouselStyles.placeholder} testID="carousel">
      <View style={CarouselStyles.placeholder} />
      <View style={CarouselStyles.lastReading}>
        <ActivityIndicator color="black" />
      </View>
      <GradientBorder x={1.0} y={1.0} />
    </View>
  )
}

Carousel.defaultProps = {
  startingIndex: 0
}

export default Carousel
