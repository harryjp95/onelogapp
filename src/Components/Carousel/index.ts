import Carousel from './Carousel'
export * from './'
export default Carousel