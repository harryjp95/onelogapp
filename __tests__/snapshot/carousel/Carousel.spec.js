import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import Carousel from '../../../src/Components/Carousel'

it('renders correctly', () => {
  const carousel = renderer.create(<Carousel symbol={'>'} handlePress={() => null}/>)

  expect(carousel).toMatchSnapshot()
})
