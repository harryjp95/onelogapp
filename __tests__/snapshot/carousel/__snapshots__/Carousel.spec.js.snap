// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 5,
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="black"
      hidesWhenStopped={true}
      size="small"
    />
  </View>
  <BVLinearGradient
    colors={
      Array [
        4293651435,
        4286611584,
        4293651435,
      ]
    }
    endPoint={
      Object {
        "x": 1,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0,
        "y": 1,
      }
    }
    style={
      Object {
        "alignItems": "center",
        "height": 0.5,
        "justifyContent": "center",
        "width": "100%",
      }
    }
  />
</View>
`;
